PWGCF/JCorran/Tasks/jflucAnalysisTask.cxx

```diff
@@ -0,0 +1,131 @@
+// or submit itself to any jurisdiction.
+/// \author Jasper Parkkila (jparkkil@cern.ch)
+/// \author Dong Jo Kim (djkim@jyu.fi)
+/// \since Sep 2022
+
+#include "Framework/runDataProcessing.h"
+#include "Framework/AnalysisTask.h"
+#include "Framework/ASoAHelpers.h"
+#include <CCDB/BasicCCDBManager.h>
+#include "Framework/RunningWorkflowInfo.h"
+#include "Framework/HistogramRegistry.h"
+
+#include "Common/DataModel/EventSelection.h"
+#include "Common/Core/TrackSelection.h"
+#include "Common/DataModel/TrackSelectionTables.h"
+#include "Common/DataModel/Centrality.h"
+#include "ReconstructionDataFormats/V0.h"
+
+#include "PWGCF/JCorran/DataModel/JCatalyst.h"
+#include "JFFlucAnalysis.h"
+
+using namespace o2;
+using namespace o2::framework;
+using namespace o2::framework::expressions;
+
+#define O2_DEFINE_CONFIGURABLE(NAME, TYPE, DEFAULT, HELP) Configurable<TYPE> NAME{#NAME, DEFAULT, HELP};
+
+class TrackIter : public JTrackIterInterface
+{
+ public:
+  TrackIter(aod::ParticleTrack::iterator _m1) : m1(_m1){};
+  ~TrackIter(){};
+  aod::ParticleTrack::iterator m1;
+  PtEtaPhiEVector v;
+  PtEtaPhiEVector& deref()
+  {
+    v.SetPt((*m1).pt());
+    v.SetEta((*m1).eta());
+    v.SetPhi((*m1).phi());
+    return v;
+  }
+  void increment()
+  {
+    ++m1;
+  }
+  bool equals(const JTrackIterInterface* prhs)
+  {
+    return m1 == static_cast<const TrackIter*>(prhs)->m1;
+  }
+};
+
+class Tracks : public TracksBase
+{
+ public:
+  Tracks(aod::ParticleTrack const* _ptracks) : ptracks(_ptracks){};
+  ~Tracks(){};
+  aod::ParticleTrack const* ptracks;
+  JTrackIter begin()
+  {
+    return JTrackIter(
+      std::unique_ptr<JTrackIterInterface>(new TrackIter(ptracks->begin())));
+  }
+  JTrackIter end()
+  {
+    /*return JTrackIter(
+      std::unique_ptr<JTrackIterInterface>(new TrackIterEnd(ptracks->end())));*/
+    return JTrackIter(
+      std::unique_ptr<JTrackIterInterface>(new TrackIter(ptracks->begin() + ptracks->size()))); // return iterator at end()
+  }
+  size_t size() const
+  {
+    return ptracks->size();
+  }
+};
+
+struct jflucAnalysisTask {
+ public:
+  ~jflucAnalysisTask()
+  {
+    delete pcf;
+  }
+
+  O2_DEFINE_CONFIGURABLE(etamin, double, 0.4, "Minimal eta for tracks");
+  O2_DEFINE_CONFIGURABLE(etamax, double, 0.8, "Maximal eta for tracks");
+
+  OutputObj<TDirectory> output{"jflucO2"};
+
+  void init(InitContext const& ic)
+  {
+    //
+    pcf = new JFFlucAnalysis("jflucAnalysis");
+    pcf->SetNumBins(sizeof(jflucCentBins) / sizeof(jflucCentBins[0]));
+    pcf->AddFlags(JFFlucAnalysis::FLUC_EBE_WEIGHTING);
+
+    output->cd();
+    pcf->UserCreateOutputObjects();
+  }
+
+  void process(soa::Join<aod::Collisions, aod::CollisionData>::iterator const& collision, aod::ParticleTrack const& tracks)
+  {
+    if (tracks.size() == 0)
+      return; // rejected event
+
+    const double fVertex[3] = {collision.posX(), collision.posY(), collision.posZ()};
+
+    Tracks tracksInt(&tracks);
```

Comment:
I guess at this point you are creating a copy of the data in the `ParticleTrack` table to your internal structure which will be used in your `UserExec` routine
You have to have in mind that this is not the aim of the O2 framework. This will probably impact the task execution time as well as its memory requirements which will penalize the execution on grid
I would suggest to adapt the `UserExec` to work directly with the tables if feasible

Reply:
Yes, this is indeed an important point. @jaelpark we can also discuss about this with a few examples how one can do it without making copies. E.g. in `correlations.cxx` are a few examples...

Reply:
In fact, the new version doesn't copy any tables or particle lists. Here ```Tracks``` is merely an iterator interface, that the correlations class ```JFFlucAnalysis``` in Core can use to access the original particle table without O2 dependencies. There's no internal structure anymore, but there's a wrapper around the native ```ParticleTrack``` iterator.